#149 update color

joshxg opened the pull request:
Fixes #149 

![image](https://user-images.githubusercontent.com/35858975/133006171-08bc1b9c-6631-4146-b75f-93e9a67c6547.png)

- I think this approach is OK. Using a special column rather than doing some tricky embedding of the color. I think the color is used sparingly enough that we don't need some sort of generic solution for it.
- I tried a few approaches to passing the `onChange` to the cell
    - Tried to add in an `onChange` into the column definition. I wasn't able to correctly type this very well - I could add the function, but couldn't pass through more generics (for the `onChange` params). Would also likely need to add sorts of handlers into the column definition? (onSubmit, onClick) which would mean they all become optional and that's a bit lame.
    - Another option is to add into the columns a `getHandler` function, which the cell can call, then the `ListView` can have a function which returns the correct callback. This is basically the same as how mobiles works. I think that it is pretty good option
    - Tried using a context and then the cell can call `const {handler} = useCellHandler(TableContext, rowKey, columnKey)` - this is basically the same as the above approach using a context rather than a `hander` callback in the column definition. I like this also.
    - Landed on a cell factory kinda pattern. I think it's pretty decent, but I *think* I like the context solution better 😁  .. this seems to work, though 🙄

joshxg commented on packages/common/src/ui/components/buttons/UnstyledIconButton.tsx:
```diff
@@ -9,13 +9,23 @@ interface ButtonProps {
     <Tooltip title={tooltip}>
-      <IconButton disabled={disabled} onClick={onClick} aria-label={tooltip}>
+      <IconButton
+        sx={{ padding: 0 }}
```
<img src="https://media4.giphy.com/media/Kc37LokcK354n7WsvN/giphy.gif?cid=5a38a5a2yayijwjyl8pw7cklkgwbn7umnlfk25jqgyp7003a&rid=giphy.gif&ct=g"/>

joshxg commented on packages/common/src/ui/components/menus/ColorMenu/ColorMenu.tsx:
```diff
@@ -62,7 +62,8 @@ export const ColorMenu: FC<ColorMenuProps> = ({
           key={hex}
-          onClick={() => {
+          onClick={e => {
+            e.stopPropagation();
```
this stops the row event from triggers. Unsure if the best place

mark-prins replied:
I think this is the right place

joshxg commented on packages/common/src/ui/layout/tables/columns/NameAndColorColumn.tsx:
```diff
@@ -0,0 +1,71 @@
+
+export const getNameAndColorColumn = <T extends Colorable>(
+  onChange: (row: T, color: Color) => void
+): Column<T> => ({
```
Was going to create an interface for `ColumnFactory` .. but wanted to see what you thought about this approach rather than the other options

joshxg commented on packages/common/src/ui/layout/tables/columns/NameAndColorColumn.tsx:
```diff
@@ -0,0 +1,71 @@
+          onClick={color => {
+            handleClose();
+            onChange(row.original, color);
+          }}
```
This is using the original data row we passed in. The `row` object has the values transformed using the `accessor` and also only contains the fields for the row (so color is technically missing in the `row` object, as the key for this row is `name`)

joshxg commented on packages/mocks/src/handlers.ts:
```diff
@@ -48,7 +49,12 @@ const upsertTransaction = graphql.mutation(
+    );
+    TransactionData[idx] = { ...TransactionData[idx], ...patch };
+
+    return response(context.data({ upsertTransaction: TransactionData[idx] }));
```
This handler was just handling new upserts. Took the rest of the code I forgot to add from the mock server to also handle actual upserts

joshxg commented on packages/transactions/src/OutboundShipment/DetailView/DetailView.tsx:
```diff
@@ -2,36 +2,17 @@ import React, { FC } from 'react';
+import { Transaction, useQueryClient } from '@openmsupply-client/common';
 
-import { getMutation, getDetailQuery } from '../../api';
+import { detailQueryFn, updateTransactionFn } from '../../api';
```
Extracted the functions in the `ListView` and `DetailView` out to the `api` file which contains all the code for the detail+list queries, delete mutation and upsert mutations as I reused the upsert mutation in the `ListView`

joshxg replied:
Think these will pretty much all go into custom hooks at some point

joshxg commented on packages/transactions/src/OutboundShipment/ListView/ListView.tsx:
```diff
@@ -63,18 +42,54 @@ export const OutboundShipmentListView: FC = () => {
+      queryClient.setQueryData(key, { ...previousData, data: previousData });
+
+      return { previousCached };
+    },
```
optimistically updating (and not handling the error case) - think there should be an easier way to do this. Likely need to do it for other things like deletion and any mutations on the `ListView` which aren't triggered by an `onSAve` event

joshxg replied:
something like: Revert + snackbar to say "Nope!"

mark-prins replied:
yeah, feels like a big manual process. if nothing else, good to extract it out if we can move this into generic hooks. logic looks good though 🤷‍♂️

joshxg reviewed (commented):
Seems sluggish. Need to investigate more, but seems like there's 3-4 re-renders of the whole thing when selecting a color 🙄

mark-prins commented on packages/transactions/src/api.ts:
```diff
@@ -46,3 +49,41 @@ export const getListQuery = (): string => gql`
+  return transaction;
+};
+
+export const updateTransactionFn = async (
```
the function names are inconsistent - in that most don't include `transaction` (presumably because they are in the `/transactions/api`) but the `updateTransactionFn` does. How about renaming to `updateFn`?

mark-prins commented on packages/transactions/src/OutboundShipment/ListView/ListView.tsx:
```diff
@@ -63,18 +42,53 @@ export const OutboundShipmentListView: FC = () => {
+      );
+      previousData[existingRowIdx] = patch;
+
+      queryClient.setQueryData(key, { ...previousData, data: previousData });
```
I'm getting confused. Shouldn't this be `...previousCached`?

mark-prins reviewed (approved):
🎉  I think it's great. covers a lot of interesting ground too, which is useful for other stuff.

I was thinking to have a separate column for color; but the name+color approach is good. likely more useful this way.

joshxg commented:
Updated the incorrect optimistic update that you caught (thanks) and renamed the func `updateFn`. Also added updates to the graphql schema for `color` that I had forgotten.
